chasm/src/main/java/org/quiltmc/chasm/internal/tree/reader/MethodNodeReader.java

```diff
@@ -430,7 +423,42 @@ private void visitTryCatchBlocks(MethodVisitor methodVisitor, MapNode codeNode)
         }
     }
 
-    private void visitLocalVariables(MethodVisitor methodVisitor) {
+    private void addLocalVariableDebugInfo() {
+        Analyzer<BasicValue> analyzer = new Analyzer<>(new BasicInterpreter());
+        Frame<BasicValue>[] frames;
+        try {
+            frames = analyzer.analyzeAndComputeMaxs(this.className, this.outputMethodNode);
+        } catch (AnalyzerException e) {
+            LOGGER.error("Failed to analyze method {}.{}{}, local variable debug info will not be present",
+                    className, outputMethodNode.name, outputMethodNode.desc);
+            return;
+        }
+        @SuppressWarnings("unchecked")
+        List<Integer>[] readabilityStarts = new List[this.outputMethodNode.maxLocals];
+        Arrays.setAll(readabilityStarts, i -> new ArrayList<>());
+        @SuppressWarnings("unchecked")
+        List<Integer>[] readabilityEnds = new List[this.outputMethodNode.maxLocals];
+        Arrays.setAll(readabilityEnds, i -> new ArrayList<>());
+        for (int insnIndex = 0; insnIndex < frames.length; insnIndex++) {
+            for (int localIndex = 0; localIndex < this.outputMethodNode.maxLocals; localIndex++) {
+                boolean isReadable = frames[insnIndex] != null
+                        && !BasicValue.UNINITIALIZED_VALUE.equals(frames[insnIndex].getLocal(localIndex));
+                boolean wasReadable = insnIndex != 0 && frames[insnIndex - 1] != null
+                        && !BasicValue.UNINITIALIZED_VALUE.equals(frames[insnIndex - 1].getLocal(localIndex));
+                if (isReadable && !wasReadable) {
+                    readabilityStarts[localIndex].add(insnIndex);
+                } else if (!isReadable && wasReadable) {
+                    readabilityEnds[localIndex].add(insnIndex);
+                }
+            }
+        }
+        for (int localIndex = 0; localIndex < this.outputMethodNode.maxLocals; localIndex++) {
+            if (readabilityEnds[localIndex].size() == readabilityStarts[localIndex].size() - 1) {
+                readabilityEnds[localIndex].add(frames.length - 1);
+            }
+            assert readabilityEnds[localIndex].size() == readabilityStarts[localIndex].size();
```

Comment:
Is assert really what we want here?

Reply:
We have decided to address this in the next PR.